browser/ui/views/brave_rewards/tip_panel_bubble_host.cc

```diff
@@ -0,0 +1,148 @@
+#include "chrome/browser/ui/views/location_bar/location_bar_view.h"
+#include "components/grit/brave_components_strings.h"
+#include "ui/gfx/geometry/rounded_corners_f.h"
+#include "ui/views/controls/webview/webview.h"
+#include "url/gurl.h"
+
+namespace brave_rewards {
+
+namespace {
+
+// The tip dialog is anchored to the center of the location bar, and must be
+// drawn with rounded bottom borders. We implement these customizations by
+// subclassing `WebUIBubbleDialogView`. This implementation is then used by
+// `TipPanelBubbleManager` when creating the dialog view.
+class TipPanelDialogView : public WebUIBubbleDialogView {
+ public:
+  TipPanelDialogView(views::View* anchor_view,
+                     BubbleContentsWrapper* contents_wrapper,
+                     const absl::optional<gfx::Rect>& anchor_rect)
+      : WebUIBubbleDialogView(anchor_view, contents_wrapper, anchor_rect) {
+    SetArrow(views::BubbleBorder::TOP_CENTER);
+  }
+
+  // views::Widget:
+  std::unique_ptr<views::NonClientFrameView> CreateNonClientFrameView(
+      views::Widget* widget) override {
+    auto frame = WebUIBubbleDialogView::CreateNonClientFrameView(widget);
+    auto* bubble_frame = static_cast<views::BubbleFrameView*>(frame.get());
+    auto* bubble_border = bubble_frame->bubble_border();
+    DCHECK(bubble_border);
+    bubble_border->SetRoundedCorners(0, 0, 16, 16);
+    return frame;
+  }
+
+  // views::View:
+  void AddedToWidget() override {
+    WebUIBubbleDialogView::AddedToWidget();
+    web_view()->holder()->SetCornerRadii(gfx::RoundedCornersF(0, 0, 16, 16));
+  }
+};
+
+// In order to use `TipPanelDialogView` as the dialog view implementation, we
+// must override the `CreateWebUIBubbleDialog` of `WebUIBubbleManager`.
+class TipPanelBubbleManager : public WebUIBubbleManager {
+ public:
+  TipPanelBubbleManager(views::View* anchor_view, Profile* profile)
+      : anchor_view_(anchor_view), profile_(profile) {}
+
+  ~TipPanelBubbleManager() override = default;
+
+  // WebUIBubbleManager:
+
+  // The persistent renderer feature is not supported for this bubble.
+  void MaybeInitPersistentRenderer() override {}
+
+  base::WeakPtr<WebUIBubbleDialogView> CreateWebUIBubbleDialog(
+      const absl::optional<gfx::Rect>& anchor) override {
+    auto contents_wrapper =
+        std::make_unique<BubbleContentsWrapperT<TipPanelUI>>(
+            GURL(kBraveTipPanelURL), profile_, IDS_BRAVE_UI_BRAVE_REWARDS);
+
+    set_bubble_using_cached_web_contents(false);
+    set_cached_contents_wrapper(std::move(contents_wrapper));
+    cached_contents_wrapper()->ReloadWebContents();
+
+    auto bubble_view = std::make_unique<TipPanelDialogView>(
+        anchor_view_, cached_contents_wrapper(), anchor);
+
+    auto weak_ptr = bubble_view->GetWeakPtr();
+    views::BubbleDialogDelegateView::CreateBubble(std::move(bubble_view));
+    return weak_ptr;
+  }
+
+ private:
+  const raw_ptr<views::View> anchor_view_;
+  const raw_ptr<Profile> profile_;
```

Comment:
Curious: what's the purpose of these `const`s?

Reply:
Yeah, saw these in a couple of places - weird! It seems that with `const` members (which are otherwise an anti-pattern) they're just trying to mimic the `const` they lose on `CreateWebUIBubbleDialog()` when overriding it.